core/modules/replica/IngestHttpSvcMod.cc

```diff
@@ -133,100 +134,199 @@ string IngestHttpSvcMod::context() const {
-
+    debug(__func__, "transactionId: "          + to_string(contrib.transactionId));
+    debug(__func__, "table: '"                 + contrib.table + "'");
+    debug(__func__, "fields_terminated_by: '"  + contrib.fieldsTerminatedBy + "'");
+    debug(__func__, "fields_enclosed_by: '"    + contrib.fieldsEnclosedBy + "'");
+    debug(__func__, "fields_escaped_by: '"     + contrib.fieldsEscapedBy + "'");
+    debug(__func__, "lines_terminated_by: '"   + contrib.linesTerminatedBy + "'");
+    debug(__func__, "chunk: "                  + to_string(contrib.chunk));
+    debug(__func__, "isOverlap: "              + string(contrib.isOverlap ? "1": "0"));
+    debug(__func__, "url: '"                   + contrib.url + "'");
+    debug(__func__, "http_method: '"           + httpMethod + "'");
+    debug(__func__, "http_data: '"             + httpData + "'");
+    debug(__func__, "http_headers.size(): "    + to_string(httpHeaders.size()));
+    debug(__func__, "expiration_timeout_sec: " + to_string(contrib.expirationTimeoutSec));
+
+    // Attempts to pass invalid transaction identifiers or tables are not recorded
+    // as transaction contributions since it's impossible to determine a context
+    // of these operations.
+
+    auto const config = serviceProvider()->config();
     auto const databaseServices = serviceProvider()->databaseServices();
-    TransactionInfo const trans = databaseServices->transaction(transactionId);
-
-    TransactionContribInfo contrib = databaseServices->beginTransactionContrib(
-        trans.id,
-        table,
-        chunk,
-        isOverlap,
-        workerInfo().name,
-        url
-    );
+
+    auto const trans = databaseServices->transaction(contrib.transactionId);
+    contrib.database = trans.database;
+
+    if (!config->databaseInfo(contrib.database).hasTable(contrib.table)) {
+        throw invalid_argument(context + "no such table '" + contrib.table + "' in database '" + contrib.database + "'.");
+    }
+
+    // Prescreen parameters of the request to ensure they're valid in the given
+    // contex. Locate and check the state of the transaction. Refuse to proceed
+    // with the request should any issues were detected.
+    bool const failed = true;
+
+    if (trans.state != TransactionInfo::STARTED) {
+        contrib.error = context + " transactionId=" + to_string(contrib.transactionId) + " is not active";
+        contrib = databaseServices->createdTransactionContrib(contrib, failed);
+        throw logic_error(contrib.error);
+    }
+
+    unique_ptr<Url> resource;
+    try {
+        resource.reset(new Url(contrib.url));
+        switch (resource->scheme()) {
+            case Url::FILE:
+            case Url::HTTP:
+            case Url::HTTPS:
+                break;
+            default:
+                throw invalid_argument(context + " unsupported url '" + contrib.url + "'");
+        }
+    } catch (exception const& ex) {
```

Comment:
Exceptions are being caught in this particular `try -> catch` loop to sanitize values of the input parameters of the REST service and report them (if any) as a failed attempt to ingest a table contribution. If the exceptions weren't caught they would be intercepted by some high-level code (the REST service processor and be reported back to a client).

The constructor of class `Url` may throw other exceptions when parsing its input. Exception `invalid_argument` is thrown for the correctly parsed `url` that is not suitable in the given context. This module only works with file-based URLs.

The algorithm won't be setting `system_error` in this `try -> catch` block because operations in the loop don't invoke any system libraries. Availability of the meaning full value for `errno` in the rest of the code depends on a particular operation. I understand the motivation behind your suggestion to generalize handling `errno`. So, I have tried to look at how to do this in the code. Unfortunately, I don't see a way to do this cleanly.

Reply:
I have simplified the code of the module by merging two `try -> catch` blocks into one.